ctx.Accepts() quality values and specificity

This issue is a follow on to #2386 and #2383 

              > i have reversed the loop for all accepts methods, so that the arguments are passed first and then the header values are checked
> 
> @efectn @sixcolors @derkan do you agree that this was the right thing to do ? the logic of express leading to the outputs implies this to me

@ReneWerner87 good job, this make the existing test cases behave as expected.

Reading the ctx.Accepts docs my understanding is that:

```
The Accepts function in the GoFiber context is used to check if the specified extensions or content types are acceptable based on the request's Accept HTTP header.

It takes one or more strings as input arguments representing the content types or extensions and returns the most preferred one based on the Accept header.
```

Therefore, if my understanding of the intent of the Accept function is correct, the following should hold true:

```
c.Request().Header.Set(HeaderAccept, "text/plain, application/json;q=0.5, text/html, */*;q=0.1")
	utils.AssertEqual(t, "txt", c.Accepts("json", "txt"))

	c.Request().Header.Set(HeaderAccept, "text/plain;charset=UTF-8;q=0.8,text/plain;charset=US-ASCII;q=0.4,text/plain;q=0.2,text/html")
	utils.AssertEqual(t, "text/plain;charset=US-ASCII", c.Accepts("text/plain", "text/plain;charset=US-ASCII"))
	utils.AssertEqual(t, "text/plain;charset=UTF-8", c.Accepts("text/plain;charset=US-ASCII", "text/plain;charset=UTF-8", "text/plain"))
	utils.AssertEqual(t, "text/html", c.Accepts("text/plain", "text/html"))
```

For example: It is possible that a browser could use the Accept header with the value text/plain;charset=UTF-8;q=0.8,text/plain;charset=US-ASCII;q=0.4,text/html. While text/html is typically assumed to have a default quality value of 1 if not specified, the browser could still explicitly specify it with a quality value of 1, as it is allowed by the HTTP specifications.

Furthermore, the browser could be indicating to the server that it prefers text/plain with a charset of UTF-8 over text/plain with a charset of US-ASCII, but it is still willing to accept both with different quality values. This could be useful in scenarios where the server has content in both character sets and wants to serve the content that is preferred by the client.

However, it is worth noting that the Accept header sent by the browser may vary depending on the user agent and its configuration. Therefore, it is important for the server to properly parse and interpret the Accept header to determine the preferred content type for the client.

## Background/Refs

The relevant RFCs for the HTTP/1.1 specification are RFC 7231 and RFC 7232.

Regarding the q parameter and its default value of 1, Section 5.3.1 of RFC 7231 states:

```
If no "*" is present in an Accept header field, then all
media types are acceptable, assuming that any media range
syntax is understood by the recipient.  If an Accept header
field is present, and if the server cannot send a response
which is acceptable according to the combined Accept field
value, then the server SHOULD send a 406 (not acceptable)
response.

[...]

Each media-range might be followed by one or more accept-
params, beginning with the "q" parameter for indicating a
relative quality factor.  The first "q" parameter (if any)
separates the media-range parameter(s) from the accept-
params.  Quality factors allow the user or user agent to
indicate the relative degree of preference for that media-
type, on a scale of 0 to 1.
```

Regarding the order of the MIME types in the Accept header, Section 5.3.2 of RFC 7231 states:

```
The order in which the content-codings are listed in the
header field is significant.  Typically, the first encoding
that is applied is the one listed as the first content-coding;
subsequent encodings are applied in the order listed.  The
order of charset values in the header field is significant,
whereas the order of language-tag values is not.  Specifically,
it does not mean that the first charset listed is preferred,
but rather that it represents the default charset if none is
otherwise specified.

An "Accept" header field can be present in a request to
indicate the formats that are acceptable for the response.
Field content negotiation is described in Section 5.3.  An
origin server that is capable of generating a response
message which is subject to content negotiation MAY vary the
response that is sent, based on the value of the Accept header
field supplied in the request.  For example, the server might
include a different Content-Type header field, or might
omit the content entirely.

The order in which the media type values are listed is
significant.  The first media type that is listed as acceptable
determines the format of the entity-body that is returned.  A
more precise or generally usable media type listed earlier in
the Accept header field SHOULD be given preference over those
that are listed later.
```

MDN on Quality Values https://developer.mozilla.org/en-US/docs/Glossary/Quality_values

```
The importance of a value is marked by the suffix ';q=' immediately followed by a value between 0 and 1 included, with up to three decimal digits, the highest value denoting the highest priority. When not present, the default value is 1.

[Examples](https://developer.mozilla.org/en-US/docs/Glossary/Quality_values#examples)
The following syntax

text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8

indicates the order of priority:

[...]

If there is no priority defined for the first two values, the order in the list is irrelevant. Nevertheless, with the same quality, more specific values have priority over less specific ones:

text/html;q=0.8,text/*;q=0.8,*/*;q=0.8

Value	Priority
text/html	0.8 (but totally specified)
text/*	0.8 (partially specified)
*/*	0.8 (not specified)

Some syntax, like the one of Accept, allow additional specifiers like text/html;level=1. These increase the specificity of the value. Their use is extremely rare.
```

_Originally posted by @sixcolors in https://github.com/gofiber/fiber/issues/2386#issuecomment-1485081978_

@ReneWerner87 

I'm working on fully cloning the express Accepts function, however I am seeing a behaviour difference from the assertions in Test_Ctx_Accepts in Fiber vs Express, specifically:

```go
	c.Request().Header.Set(HeaderAccept, "text/html, application/json")
	utils.AssertEqual(t, "text/*", c.Accepts("text/*"))
```

Results in:

```bash
--- FAIL: Test_Ctx_Accepts (0.01s)
    /Users/sixcolors/Documents/GitHub/fiber/ctx_test.go:58: 
        Test:       Test_Ctx_Accepts
        Trace:      ctx_test.go:58
        Expect:     text/*     (string)
        Result:                (string)
FAIL
FAIL	github.com/gofiber/fiber/v2	0.029s
FAIL
```

The express / js function in the following sample

```js
const express = require('express');
var accepts = require('accepts')
var Negotiator = require('negotiator')

const app = express();
const port = 3000;

const availableMediaTypes = ['text/*'];

app.use(express.json());

app.get('/', (req, res) => {
    negotiator = new Negotiator(req);
    console.log(negotiator.mediaTypes(availableMediaTypes));

    var accept = accepts(req)

    console.log(accept.types(availableMediaTypes) ? 'accepts' : 'does not accept');

    req.accepts(availableMediaTypes) ? console.log('express accepts') : console.log('express does not accept');
    res.send('Hello, world!');
});

app.post('/data', (req, res) => {
    if (req.accepts('json')) {
        res.json({ message: 'Received JSON data' });
    } else if (req.accepts('xml')) {
        res.send('<message>Received XML data</message>');
    } else {
        res.status(406).send('Not Acceptable');
    }
});

app.listen(port, () => {
    console.log(`Server listening on port ${port}`);
});
```

```bash
curl -v -H'Accept: text/html, application/json' http://localhost:3000/
```

Gives the console output:

```console
[]
does not accept
express does not accept
```

--- a/ctx_test.go
+++ b/ctx_test.go
@@ -45,7 +45,10 @@ func Test_Ctx_Accepts(t *testing.T) {
 	utils.AssertEqual(t, "", c.Accepts())
 	utils.AssertEqual(t, ".xml", c.Accepts(".xml"))
 	utils.AssertEqual(t, "", c.Accepts(".john"))
-	utils.AssertEqual(t, "application/xhtml+xml", c.Accepts("application/xml", "application/xml+rss", "application/yml", "application/xhtml+xml"))
+	utils.AssertEqual(t, "application/xhtml+xml", c.Accepts("application/xml", "application/xml+rss", "application/yml", "application/xhtml+xml"), "must use client-preferred mime type")
+
+	c.Request().Header.Set(HeaderAccept, "application/json, text/plain, */*;q=0")
+	utils.AssertEqual(t, "", c.Accepts("html"), "must treat */*;q=0 as not acceptable")
 
 	c.Request().Header.Set(HeaderAccept, "text/*, application/json")
 	utils.AssertEqual(t, "html", c.Accepts("html"))

--- a/ctx_test.go
+++ b/ctx_test.go
@@ -45,7 +45,7 @@ func Test_Ctx_Accepts(t *testing.T) {
 	utils.AssertEqual(t, "", c.Accepts())
 	utils.AssertEqual(t, ".xml", c.Accepts(".xml"))
 	utils.AssertEqual(t, "", c.Accepts(".john"))
-	utils.AssertEqual(t, "application/xhtml+xml", c.Accepts("application/xml", "application/xml+rss", "application/yml", "application/xhtml+xml"), "must use client-preferred mime type")
+	utils.AssertEqual(t, "application/xhtml+xml", c.Accepts("application/xml", "application/xml+rss", "application/yaml", "application/xhtml+xml"), "must use client-preferred mime type")
 
 	c.Request().Header.Set(HeaderAccept, "application/json, text/plain, */*;q=0")
 	utils.AssertEqual(t, "", c.Accepts("html"), "must treat */*;q=0 as not acceptable")

--- a/ctx_test.go
+++ b/ctx_test.go
@@ -45,6 +45,7 @@ func Test_Ctx_Accepts(t *testing.T) {
 	utils.AssertEqual(t, "", c.Accepts())
 	utils.AssertEqual(t, ".xml", c.Accepts(".xml"))
 	utils.AssertEqual(t, "", c.Accepts(".john"))
+	utils.AssertEqual(t, "application/xhtml+xml", c.Accepts("application/xml", "application/xml+rss", "application/yml", "application/xhtml+xml"))
 
 	c.Request().Header.Set(HeaderAccept, "text/*, application/json")
 	utils.AssertEqual(t, "html", c.Accepts("html"))

--- a/docs/api/ctx.md
+++ b/docs/api/ctx.md
@@ -24,13 +24,16 @@ func (c *Ctx) AcceptsLanguages(offers ...string) string
 ```
 
 ```go title="Example"
-// Accept: text/*, application/json
+// Accept: text/html, application/json; q=0.8, text/plain; q=0.5; charset="utf-8"
 
 app.Get("/", func(c *fiber.Ctx) error {
   c.Accepts("html")             // "html"
   c.Accepts("text/html")        // "text/html"
   c.Accepts("json", "text")     // "json"
   c.Accepts("application/json") // "application/json"
+  c.Accepts("text/plain", "application/json") // "application/json", due to quality
+  c.Accepts("text/plain", "application/json") // "application/json", due to specificity
+  c.Accepts("text/html", "application/json") // "text/html", due to first match
   c.Accepts("image/png")        // ""
   c.Accepts("png")              // ""
   // ...

--- a/docs/api/ctx.md
+++ b/docs/api/ctx.md
@@ -39,7 +39,7 @@ app.Get("/", func(c *fiber.Ctx) error {
 ```
 
 ```go title="Example 2"
-// Accept: text/html, text/*, application/json, */*; q=0"
+// Accept: text/html, text/*, application/json, */*; q=0
 
 app.Get("/", func(c *fiber.Ctx) error {
   c.Accepts("text/plain", "application/json") // "application/json", due to specificity

--- a/docs/api/ctx.md
+++ b/docs/api/ctx.md
@@ -32,10 +32,19 @@ app.Get("/", func(c *fiber.Ctx) error {
   c.Accepts("json", "text")     // "json"
   c.Accepts("application/json") // "application/json"
   c.Accepts("text/plain", "application/json") // "application/json", due to quality
-  c.Accepts("text/plain", "application/json") // "application/json", due to specificity
-  c.Accepts("text/html", "application/json") // "text/html", due to first match
   c.Accepts("image/png")        // ""
   c.Accepts("png")              // ""
+  // ...
+})
+```
+
+```go title="Example 2"
+// Accept: text/html, text/*, application/json, */*; q=0"
+
+app.Get("/", func(c *fiber.Ctx) error {
+  c.Accepts("text/plain", "application/json") // "application/json", due to specificity
+  c.Accepts("application/json", "text/html") // "text/html", due to first match
+  c.Accepts("image/png")        // "", due to */* without q factor 0 is Not Acceptable
   // ...
 })
 ```

--- a/helpers.go
+++ b/helpers.go
@@ -295,7 +295,9 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
 			factor := utils.Trim(spec[factorSign+1:], ' ')
 			if strings.HasPrefix(factor, "q=") {
-				quality, _ = strconv.ParseFloat(factor[2:], 64)
+				if q, err := strconv.ParseFloat(factor[2:], 64); err != nil {
+					quality = q
+				}
 			}
 			spec = spec[:factorSign]
 		}

--- a/helpers.go
+++ b/helpers.go
@@ -15,6 +15,7 @@ import (
 	"os"
 	"path/filepath"
 	"reflect"
+	"sort"
 	"strconv"
 	"strings"
 	"time"
@@ -336,7 +337,14 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 	}
 
 	// Sort accepted types by quality and specificity
-	if len(acceptedTypes) > 1 {
+	if len(acceptedTypes) > 20 {
+		sort.SliceStable(acceptedTypes, func(i, j int) bool {
+			if acceptedTypes[i].quality == acceptedTypes[j].quality {
+				return acceptedTypes[i].specificity > acceptedTypes[j].specificity
+			}
+			return acceptedTypes[i].quality > acceptedTypes[j].quality
+		})
+	} else if len(acceptedTypes) > 1 {
 		// Insertion sort
 		changes := false
 		for i := 1; i < len(acceptedTypes); i++ {

--- a/helpers.go
+++ b/helpers.go
@@ -15,7 +15,6 @@ import (
 	"os"
 	"path/filepath"
 	"reflect"
-	"sort"
 	"strconv"
 	"strings"
 	"time"
@@ -337,14 +336,7 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 	}
 
 	// Sort accepted types by quality and specificity
-	if len(acceptedTypes) > 6 {
-		sort.SliceStable(acceptedTypes, func(i, j int) bool {
-			if acceptedTypes[i].quality == acceptedTypes[j].quality {
-				return acceptedTypes[i].specificity > acceptedTypes[j].specificity
-			}
-			return acceptedTypes[i].quality > acceptedTypes[j].quality
-		})
-	} else if len(acceptedTypes) > 1 {
+	if len(acceptedTypes) > 1 {
 		// Insertion sort
 		changes := false
 		for i := 1; i < len(acceptedTypes); i++ {

--- a/helpers.go
+++ b/helpers.go
@@ -337,16 +337,24 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 
 	// Sort accepted types by quality and specificity
 	if len(acceptedTypes) > 1 {
-		// Insertion sort
 		for i := 1; i < len(acceptedTypes); i++ {
-			for j := i; j > 0; j-- {
-				if acceptedTypes[j-1].quality < acceptedTypes[j].quality {
-					acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
-				} else if acceptedTypes[j-1].quality == acceptedTypes[j].quality && acceptedTypes[j-1].specificity < acceptedTypes[j].specificity {
-					acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
+			swap := false
+			lo, hi := 0, i-1
+			for lo <= hi {
+				mid := (lo + hi) / 2
+				if acceptedTypes[i].quality > acceptedTypes[mid].quality ||
+					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity > acceptedTypes[mid].specificity) {
+					lo = mid + 1
 				} else {
-					break
+					hi = mid - 1
 				}
+			}
+			for j := i; j > lo; j-- {
+				acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
+				swap = true
+			}
+			if !swap {
+				break
 			}
 		}
 	}

--- a/helpers.go
+++ b/helpers.go
@@ -15,7 +15,6 @@ import (
 	"os"
 	"path/filepath"
 	"reflect"
-	"strconv"
 	"strings"
 	"time"
 	"unsafe"
@@ -298,7 +297,7 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
 			factor := utils.Trim(spec[factorSign+1:], ' ')
 			if strings.HasPrefix(factor, "q=") {
-				if q, err := strconv.ParseFloat(factor[2:], 64); err == nil {
+				if q, err := fasthttp.ParseUfloat([]byte(factor[2:])); err == nil {
 					quality = q
 				}
 			}

--- a/helpers.go
+++ b/helpers.go
@@ -272,13 +272,16 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		spec        string
 		quality     float64
 		specificity int
+		order       int
 	}
 
 	// Parse header and get accepted types with their quality and specificity
 	// See: https://www.rfc-editor.org/rfc/rfc9110#name-content-negotiation-fields
-	spec, commaPos := "", 0
+	spec, commaPos, order := "", 0, 0
 	acceptedTypes := make([]acceptedType, 0, 20)
 	for len(header) > 0 {
+		order++
+
 		// Skip spaces
 		header = utils.TrimLeft(header, ' ')
 
@@ -295,7 +298,7 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
 			factor := utils.Trim(spec[factorSign+1:], ' ')
 			if strings.HasPrefix(factor, "q=") {
-				if q, err := strconv.ParseFloat(factor[2:], 64); err != nil {
+				if q, err := strconv.ParseFloat(factor[2:], 64); err == nil {
 					quality = q
 				}
 			}
@@ -327,7 +330,7 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		}
 
 		// Add to accepted types
-		acceptedTypes = append(acceptedTypes, acceptedType{spec, quality, specificity})
+		acceptedTypes = append(acceptedTypes, acceptedType{spec, quality, specificity, order})
 
 		// Next
 		if commaPos != -1 {
@@ -337,15 +340,16 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		}
 	}
 
-	// Sort accepted types by quality and specificity
+	// Sort accepted types by quality and specificity, preserving order of equal elements
 	if len(acceptedTypes) > 1 {
 		for i := 1; i < len(acceptedTypes); i++ {
 			swap := false
 			lo, hi := 0, i-1
 			for lo <= hi {
 				mid := (lo + hi) / 2
-				if acceptedTypes[i].quality > acceptedTypes[mid].quality ||
-					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity > acceptedTypes[mid].specificity) {
+				if acceptedTypes[i].quality < acceptedTypes[mid].quality ||
+					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity < acceptedTypes[mid].specificity) ||
+					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity == acceptedTypes[mid].specificity && acceptedTypes[i].order > acceptedTypes[mid].order) {
 					lo = mid + 1
 				} else {
 					hi = mid - 1

--- a/helpers.go
+++ b/helpers.go
@@ -300,7 +300,7 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 			spec = spec[:factorSign]
 		}
 
-		if quality == 0 {
+		if quality == 0.0 {
 			// Skip this spec
 			if commaPos != -1 {
 				header = header[commaPos+1:]

--- a/helpers.go
+++ b/helpers.go
@@ -25,6 +25,16 @@ import (
 	"github.com/valyala/fasthttp"
 )
 
+// acceptType is a struct that holds the parsed value of an Accept header
+// along with quality, specificity, and order.
+// used for sorting accept headers.
+type acceptedType struct {
+	spec        string
+	quality     float64
+	specificity int
+	order       int
+}
+
 // getTLSConfig returns a net listener's tls config
 func getTLSConfig(ln net.Listener) *tls.Config {
 	// Get listener type
@@ -261,17 +271,14 @@ func acceptsOfferType(spec, offerType string) bool {
 
 // getOffer return valid offer for header negotiation
 func getOffer(header string, isAccepted func(spec, offer string) bool, offers ...string) string {
+	if isAccepted == nil {
+		return ""
+	}
 	if len(offers) == 0 {
 		return ""
-	} else if header == "" {
-		return offers[0]
 	}
-
-	type acceptedType struct {
-		spec        string
-		quality     float64
-		specificity int
-		order       int
+	if header == "" {
+		return offers[0]
 	}
 
 	// Parse header and get accepted types with their quality and specificity
@@ -339,29 +346,9 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		}
 	}
 
-	// Sort accepted types by quality and specificity, preserving order of equal elements
 	if len(acceptedTypes) > 1 {
-		for i := 1; i < len(acceptedTypes); i++ {
-			swap := false
-			lo, hi := 0, i-1
-			for lo <= hi {
-				mid := (lo + hi) / 2
-				if acceptedTypes[i].quality < acceptedTypes[mid].quality ||
-					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity < acceptedTypes[mid].specificity) ||
-					(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity == acceptedTypes[mid].specificity && acceptedTypes[i].order > acceptedTypes[mid].order) {
-					lo = mid + 1
-				} else {
-					hi = mid - 1
-				}
-			}
-			for j := i; j > lo; j-- {
-				acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
-				swap = true
-			}
-			if !swap {
-				break
-			}
-		}
+		// Sort accepted types by quality and specificity, preserving order of equal elements
+		sortAcceptedTypes(&acceptedTypes)
 	}
 
 	// Find the first offer that matches the accepted types
@@ -377,6 +364,35 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 	}
 
 	return ""
+}
+
+// sortAcceptedTypes sorts accepted types by quality and specificity, preserving order of equal elements
+//
+// Parameters are not supported, they are ignored when sorting by specificity.
+//
+// See: https://www.rfc-editor.org/rfc/rfc9110#name-content-negotiation-fields
+func sortAcceptedTypes(at *[]acceptedType) {
+	if at == nil || len(*at) < 2 {
+		return
+	}
+	acceptedTypes := *at
+
+	for i := 1; i < len(acceptedTypes); i++ {
+		lo, hi := 0, i-1
+		for lo <= hi {
+			mid := (lo + hi) / 2
+			if acceptedTypes[i].quality < acceptedTypes[mid].quality ||
+				(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity < acceptedTypes[mid].specificity) ||
+				(acceptedTypes[i].quality == acceptedTypes[mid].quality && acceptedTypes[i].specificity == acceptedTypes[mid].specificity && acceptedTypes[i].order > acceptedTypes[mid].order) {
+				lo = mid + 1
+			} else {
+				hi = mid - 1
+			}
+		}
+		for j := i; j > lo; j-- {
+			acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
+		}
+	}
 }
 
 func matchEtag(s, etag string) bool {

--- a/helpers.go
+++ b/helpers.go
@@ -301,7 +301,7 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
 			factor := utils.Trim(spec[factorSign+1:], ' ')
 			if strings.HasPrefix(factor, "q=") {
-				if q, err := fasthttp.ParseUfloat([]byte(factor[2:])); err == nil {
+				if q, err := fasthttp.ParseUfloat(utils.UnsafeBytes(factor[2:])); err == nil {
 					quality = q
 				}
 			}

--- a/helpers.go
+++ b/helpers.go
@@ -275,7 +275,8 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		specificity int
 	}
 
-	// Parse header
+	// Parse header and get accepted types with their quality and specificity
+	// See: https://www.rfc-editor.org/rfc/rfc9110#name-content-negotiation-fields
 	spec, commaPos := "", 0
 	acceptedTypes := make([]acceptedType, 0)
 	for len(header) > 0 {
@@ -300,8 +301,9 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 			spec = spec[:factorSign]
 		}
 
+		// Skip if quality is 0.0
+		// See: https://www.rfc-editor.org/rfc/rfc9110#quality.values
 		if quality == 0.0 {
-			// Skip this spec
 			if commaPos != -1 {
 				header = header[commaPos+1:]
 			} else {

--- a/helpers.go
+++ b/helpers.go
@@ -271,9 +271,6 @@ func acceptsOfferType(spec, offerType string) bool {
 
 // getOffer return valid offer for header negotiation
 func getOffer(header string, isAccepted func(spec, offer string) bool, offers ...string) string {
-	if isAccepted == nil {
-		return ""
-	}
 	if len(offers) == 0 {
 		return ""
 	}

--- a/helpers.go
+++ b/helpers.go
@@ -300,6 +300,16 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 			spec = spec[:factorSign]
 		}
 
+		if quality == 0 {
+			// Skip this spec
+			if commaPos != -1 {
+				header = header[commaPos+1:]
+			} else {
+				break
+			}
+			continue
+		}
+
 		// Get specificity
 		specificity := 0
 		// check for wildcard this could be a mime */* or a wildcard character *

--- a/helpers.go
+++ b/helpers.go
@@ -15,6 +15,8 @@ import (
 	"os"
 	"path/filepath"
 	"reflect"
+	"sort"
+	"strconv"
 	"strings"
 	"time"
 	"unsafe"
@@ -267,29 +269,97 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 		return offers[0]
 	}
 
-	for _, offer := range offers {
-		if len(offer) == 0 {
-			continue
+	type acceptedType struct {
+		spec        string
+		quality     float64
+		specificity int
+	}
+
+	// Parse header
+	spec, commaPos := "", 0
+	acceptedTypes := make([]acceptedType, 0)
+	for len(header) > 0 {
+		// Skip spaces
+		header = utils.TrimLeft(header, ' ')
+
+		// Get spec
+		commaPos = strings.IndexByte(header, ',')
+		if commaPos != -1 {
+			spec = utils.Trim(header[:commaPos], ' ')
+		} else {
+			spec = utils.TrimLeft(header, ' ')
 		}
-		spec, commaPos := "", 0
-		for len(header) > 0 && commaPos != -1 {
-			commaPos = strings.IndexByte(header, ',')
-			if commaPos != -1 {
-				spec = utils.Trim(header[:commaPos], ' ')
-			} else {
-				spec = utils.TrimLeft(header, ' ')
-			}
-			if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
-				spec = spec[:factorSign]
-			}
 
-			// isAccepted if the current offer is accepted
-			if isAccepted(spec, offer) {
+		// Get quality
+		quality := 1.0
+		if factorSign := strings.IndexByte(spec, ';'); factorSign != -1 {
+			factor := utils.Trim(spec[factorSign+1:], ' ')
+			if strings.HasPrefix(factor, "q=") {
+				quality, _ = strconv.ParseFloat(factor[2:], 64)
+			}
+			spec = spec[:factorSign]
+		}
+
+		// Get specificity
+		specificity := 0
+		// check for wildcard this could be a mime */* or a wildcard character *
+		if spec == "*/*" || spec == "*" {
+			specificity = 1
+		} else if strings.HasSuffix(spec, "/*") {
+			specificity = 2
+		} else if strings.IndexByte(spec, '/') != -1 {
+			specificity = 3
+		} else {
+			specificity = 4
+		}
+
+		// Add to accepted types
+		acceptedTypes = append(acceptedTypes, acceptedType{spec, quality, specificity})
+
+		// Next
+		if commaPos != -1 {
+			header = header[commaPos+1:]
+		} else {
+			break
+		}
+	}
+
+	// Sort accepted types by quality and specificity
+	if len(acceptedTypes) > 6 {
+		sort.SliceStable(acceptedTypes, func(i, j int) bool {
+			if acceptedTypes[i].quality == acceptedTypes[j].quality {
+				return acceptedTypes[i].specificity > acceptedTypes[j].specificity
+			}
+			return acceptedTypes[i].quality > acceptedTypes[j].quality
+		})
+	} else if len(acceptedTypes) > 1 {
+		// Insertion sort
+		changes := false
+		for i := 1; i < len(acceptedTypes); i++ {
+			for j := i; j > 0; j-- {
+				if acceptedTypes[j-1].quality < acceptedTypes[j].quality {
+					acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
+					changes = true
+				} else if acceptedTypes[j-1].quality == acceptedTypes[j].quality && acceptedTypes[j-1].specificity < acceptedTypes[j].specificity {
+					acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
+					changes = true
+				}
+
+				if !changes {
+					break
+				}
+			}
+		}
+	}
+
+	// Find the first offer that matches the accepted types
+	for _, acceptedType := range acceptedTypes {
+		for _, offer := range offers {
+			if len(offer) == 0 {
+				continue
+			}
+			if isAccepted(acceptedType.spec, offer) {
 				return offer
-			}
-
-			if commaPos != -1 {
-				header = header[commaPos+1:]
 			}
 		}
 	}

--- a/helpers.go
+++ b/helpers.go
@@ -15,7 +15,6 @@ import (
 	"os"
 	"path/filepath"
 	"reflect"
-	"sort"
 	"strconv"
 	"strings"
 	"time"
@@ -278,7 +277,7 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 	// Parse header and get accepted types with their quality and specificity
 	// See: https://www.rfc-editor.org/rfc/rfc9110#name-content-negotiation-fields
 	spec, commaPos := "", 0
-	acceptedTypes := make([]acceptedType, 0)
+	acceptedTypes := make([]acceptedType, 0, 20)
 	for len(header) > 0 {
 		// Skip spaces
 		header = utils.TrimLeft(header, ' ')
@@ -337,27 +336,15 @@ func getOffer(header string, isAccepted func(spec, offer string) bool, offers ..
 	}
 
 	// Sort accepted types by quality and specificity
-	if len(acceptedTypes) > 20 {
-		sort.SliceStable(acceptedTypes, func(i, j int) bool {
-			if acceptedTypes[i].quality == acceptedTypes[j].quality {
-				return acceptedTypes[i].specificity > acceptedTypes[j].specificity
-			}
-			return acceptedTypes[i].quality > acceptedTypes[j].quality
-		})
-	} else if len(acceptedTypes) > 1 {
+	if len(acceptedTypes) > 1 {
 		// Insertion sort
-		changes := false
 		for i := 1; i < len(acceptedTypes); i++ {
 			for j := i; j > 0; j-- {
 				if acceptedTypes[j-1].quality < acceptedTypes[j].quality {
 					acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
-					changes = true
 				} else if acceptedTypes[j-1].quality == acceptedTypes[j].quality && acceptedTypes[j-1].specificity < acceptedTypes[j].specificity {
 					acceptedTypes[j-1], acceptedTypes[j] = acceptedTypes[j], acceptedTypes[j-1]
-					changes = true
-				}
-
-				if !changes {
+				} else {
 					break
 				}
 			}

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -221,7 +221,7 @@ func Test_Utils_Parse_Address(t *testing.T) {
 	}
 }
 
-func Test_GetOffer(t *testing.T) {
+func Test_Utils_GetOffer(t *testing.T) {
 	t.Parallel()
 	utils.AssertEqual(t, "", getOffer("hello", acceptsOffer))
 	utils.AssertEqual(t, "1", getOffer("", acceptsOffer, "1"))

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -88,11 +88,13 @@ func Test_Utils_GetOffer(t *testing.T) {
 func Benchmark_Utils_GetOffer(b *testing.B) {
 	headers := []string{
 		"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
+		"application/json",
 		"utf-8, iso-8859-1;q=0.5",
 		"gzip, deflate",
 	}
 	offers := [][]string{
-		{"application/xml", "application/json"},
+		{"text/html", "application/xml", "application/xml+xhtml"},
+		{"application/json"},
 		{"utf-8"},
 		{"deflate"},
 	}
@@ -101,6 +103,37 @@ func Benchmark_Utils_GetOffer(b *testing.B) {
 			getOffer(header, acceptsOfferType, offers[i]...)
 		}
 	}
+}
+
+func Test_Utils_SortAcceptedTypes(t *testing.T) {
+	t.Parallel()
+	acceptedTypes := []acceptedType{
+		{spec: "text/html", quality: 1, specificity: 3, order: 0},
+		{spec: "text/*", quality: 0.5, specificity: 2, order: 1},
+		{spec: "*/*", quality: 0.1, specificity: 1, order: 2},
+		{spec: "application/json", quality: 0.999, specificity: 3, order: 3},
+		{spec: "application/xml", quality: 1, specificity: 3, order: 4},
+		{spec: "application/pdf", quality: 1, specificity: 3, order: 5},
+		{spec: "image/png", quality: 1, specificity: 3, order: 6},
+		{spec: "image/jpeg", quality: 1, specificity: 3, order: 7},
+		{spec: "image/*", quality: 1, specificity: 2, order: 8},
+		{spec: "image/gif", quality: 1, specificity: 3, order: 9},
+		{spec: "text/plain", quality: 1, specificity: 3, order: 10},
+	}
+	sortAcceptedTypes(&acceptedTypes)
+	utils.AssertEqual(t, acceptedTypes, []acceptedType{
+		{spec: "text/html", quality: 1, specificity: 3, order: 0},
+		{spec: "application/xml", quality: 1, specificity: 3, order: 4},
+		{spec: "application/pdf", quality: 1, specificity: 3, order: 5},
+		{spec: "image/png", quality: 1, specificity: 3, order: 6},
+		{spec: "image/jpeg", quality: 1, specificity: 3, order: 7},
+		{spec: "image/gif", quality: 1, specificity: 3, order: 9},
+		{spec: "text/plain", quality: 1, specificity: 3, order: 10},
+		{spec: "image/*", quality: 1, specificity: 2, order: 8},
+		{spec: "application/json", quality: 0.999, specificity: 3, order: 3},
+		{spec: "text/*", quality: 0.5, specificity: 2, order: 1},
+		{spec: "*/*", quality: 0.1, specificity: 1, order: 2},
+	})
 }
 
 // go test -v -run=^$ -bench=Benchmark_Utils_SortAcceptedTypes_Sorted -benchmem -count=4
@@ -124,7 +157,7 @@ func Benchmark_Utils_SortAcceptedTypes_Unsorted(b *testing.B) {
 		acceptedTypes[0] = acceptedType{spec: "text/html", quality: 1, specificity: 3, order: 0}
 		acceptedTypes[1] = acceptedType{spec: "text/*", quality: 0.5, specificity: 2, order: 1}
 		acceptedTypes[2] = acceptedType{spec: "*/*", quality: 0.1, specificity: 1, order: 2}
-		acceptedTypes[3] = acceptedType{spec: "application/json", quality: 1, specificity: 3, order: 3}
+		acceptedTypes[3] = acceptedType{spec: "application/json", quality: 0.999, specificity: 3, order: 3}
 		acceptedTypes[4] = acceptedType{spec: "application/xml", quality: 1, specificity: 3, order: 4}
 		acceptedTypes[5] = acceptedType{spec: "application/pdf", quality: 1, specificity: 3, order: 5}
 		acceptedTypes[6] = acceptedType{spec: "image/png", quality: 1, specificity: 3, order: 6}
@@ -134,17 +167,19 @@ func Benchmark_Utils_SortAcceptedTypes_Unsorted(b *testing.B) {
 		acceptedTypes[10] = acceptedType{spec: "text/plain", quality: 1, specificity: 3, order: 10}
 		sortAcceptedTypes(&acceptedTypes)
 	}
-	utils.AssertEqual(b, "text/html", acceptedTypes[0].spec)
-	utils.AssertEqual(b, "application/json", acceptedTypes[1].spec)
-	utils.AssertEqual(b, "application/xml", acceptedTypes[2].spec)
-	utils.AssertEqual(b, "application/pdf", acceptedTypes[3].spec)
-	utils.AssertEqual(b, "image/png", acceptedTypes[4].spec)
-	utils.AssertEqual(b, "image/jpeg", acceptedTypes[5].spec)
-	utils.AssertEqual(b, "image/gif", acceptedTypes[6].spec)
-	utils.AssertEqual(b, "text/plain", acceptedTypes[7].spec)
-	utils.AssertEqual(b, "image/*", acceptedTypes[8].spec)
-	utils.AssertEqual(b, "text/*", acceptedTypes[9].spec)
-	utils.AssertEqual(b, "*/*", acceptedTypes[10].spec)
+	utils.AssertEqual(b, acceptedTypes, []acceptedType{
+		{spec: "text/html", quality: 1, specificity: 3, order: 0},
+		{spec: "application/xml", quality: 1, specificity: 3, order: 4},
+		{spec: "application/pdf", quality: 1, specificity: 3, order: 5},
+		{spec: "image/png", quality: 1, specificity: 3, order: 6},
+		{spec: "image/jpeg", quality: 1, specificity: 3, order: 7},
+		{spec: "image/gif", quality: 1, specificity: 3, order: 9},
+		{spec: "text/plain", quality: 1, specificity: 3, order: 10},
+		{spec: "image/*", quality: 1, specificity: 2, order: 8},
+		{spec: "application/json", quality: 0.999, specificity: 3, order: 3},
+		{spec: "text/*", quality: 0.5, specificity: 2, order: 1},
+		{spec: "*/*", quality: 0.1, specificity: 1, order: 2},
+	})
 }
 
 // go test -v -run=^$ -bench=Benchmark_App_ETag -benchmem -count=4

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -221,7 +221,7 @@ func Test_Utils_Parse_Address(t *testing.T) {
 	}
 }
 
-func Test_Utils_GetOffset(t *testing.T) {
+func Test_Utils_GetOffer(t *testing.T) {
 	t.Parallel()
 	utils.AssertEqual(t, "", getOffer("hello", acceptsOffer))
 	utils.AssertEqual(t, "1", getOffer("", acceptsOffer, "1"))

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -62,6 +62,91 @@ func Test_Utils_ETag(t *testing.T) {
 	})
 }
 
+func Test_Utils_GetOffer(t *testing.T) {
+	t.Parallel()
+	utils.AssertEqual(t, "", getOffer("hello", acceptsOffer))
+	utils.AssertEqual(t, "1", getOffer("", acceptsOffer, "1"))
+	utils.AssertEqual(t, "", getOffer("2", acceptsOffer, "1"))
+
+	utils.AssertEqual(t, "", getOffer("", acceptsOfferType))
+	utils.AssertEqual(t, "", getOffer("text/html", acceptsOfferType))
+	utils.AssertEqual(t, "", getOffer("text/html", acceptsOfferType, "application/json"))
+	utils.AssertEqual(t, "", getOffer("text/html;q=0", acceptsOfferType, "text/html"))
+	utils.AssertEqual(t, "application/xml", getOffer("text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", acceptsOfferType, "application/xml", "application/json"))
+	utils.AssertEqual(t, "text/html", getOffer("text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", acceptsOfferType, "text/html"))
+	utils.AssertEqual(t, "application/pdf", getOffer("text/plain;q=0,application/pdf;q=0.9,*/*;q=0.000", acceptsOfferType, "application/pdf", "application/json"))
+
+	utils.AssertEqual(t, "", getOffer("utf-8, iso-8859-1;q=0.5", acceptsOffer))
+	utils.AssertEqual(t, "", getOffer("utf-8, iso-8859-1;q=0.5", acceptsOffer, "ascii"))
+	utils.AssertEqual(t, "utf-8", getOffer("utf-8, iso-8859-1;q=0.5", acceptsOffer, "utf-8"))
+	utils.AssertEqual(t, "iso-8859-1", getOffer("utf-8;q=0, iso-8859-1;q=0.5", acceptsOffer, "utf-8", "iso-8859-1"))
+
+	utils.AssertEqual(t, "deflate", getOffer("gzip, deflate", acceptsOffer, "deflate"))
+	utils.AssertEqual(t, "", getOffer("gzip, deflate;q=0", acceptsOffer, "deflate"))
+}
+
+func Benchmark_Utils_GetOffer(b *testing.B) {
+	headers := []string{
+		"text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
+		"utf-8, iso-8859-1;q=0.5",
+		"gzip, deflate",
+	}
+	offers := [][]string{
+		{"application/xml", "application/json"},
+		{"utf-8"},
+		{"deflate"},
+	}
+	for n := 0; n < b.N; n++ {
+		for i, header := range headers {
+			getOffer(header, acceptsOfferType, offers[i]...)
+		}
+	}
+}
+
+// go test -v -run=^$ -bench=Benchmark_Utils_SortAcceptedTypes_Sorted -benchmem -count=4
+func Benchmark_Utils_SortAcceptedTypes_Sorted(b *testing.B) {
+	acceptedTypes := make([]acceptedType, 3)
+	for n := 0; n < b.N; n++ {
+		acceptedTypes[0] = acceptedType{spec: "text/html", quality: 1, specificity: 1, order: 0}
+		acceptedTypes[1] = acceptedType{spec: "text/*", quality: 0.5, specificity: 1, order: 1}
+		acceptedTypes[2] = acceptedType{spec: "*/*", quality: 0.1, specificity: 1, order: 2}
+		sortAcceptedTypes(&acceptedTypes)
+	}
+	utils.AssertEqual(b, "text/html", acceptedTypes[0].spec)
+	utils.AssertEqual(b, "text/*", acceptedTypes[1].spec)
+	utils.AssertEqual(b, "*/*", acceptedTypes[2].spec)
+}
+
+// go test -v -run=^$ -bench=Benchmark_Utils_SortAcceptedTypes_Unsorted -benchmem -count=4
+func Benchmark_Utils_SortAcceptedTypes_Unsorted(b *testing.B) {
+	acceptedTypes := make([]acceptedType, 11)
+	for n := 0; n < b.N; n++ {
+		acceptedTypes[0] = acceptedType{spec: "text/html", quality: 1, specificity: 3, order: 0}
+		acceptedTypes[1] = acceptedType{spec: "text/*", quality: 0.5, specificity: 2, order: 1}
+		acceptedTypes[2] = acceptedType{spec: "*/*", quality: 0.1, specificity: 1, order: 2}
+		acceptedTypes[3] = acceptedType{spec: "application/json", quality: 1, specificity: 3, order: 3}
+		acceptedTypes[4] = acceptedType{spec: "application/xml", quality: 1, specificity: 3, order: 4}
+		acceptedTypes[5] = acceptedType{spec: "application/pdf", quality: 1, specificity: 3, order: 5}
+		acceptedTypes[6] = acceptedType{spec: "image/png", quality: 1, specificity: 3, order: 6}
+		acceptedTypes[7] = acceptedType{spec: "image/jpeg", quality: 1, specificity: 3, order: 7}
+		acceptedTypes[8] = acceptedType{spec: "image/*", quality: 1, specificity: 2, order: 8}
+		acceptedTypes[9] = acceptedType{spec: "image/gif", quality: 1, specificity: 3, order: 9}
+		acceptedTypes[10] = acceptedType{spec: "text/plain", quality: 1, specificity: 3, order: 10}
+		sortAcceptedTypes(&acceptedTypes)
+	}
+	utils.AssertEqual(b, "text/html", acceptedTypes[0].spec)
+	utils.AssertEqual(b, "application/json", acceptedTypes[1].spec)
+	utils.AssertEqual(b, "application/xml", acceptedTypes[2].spec)
+	utils.AssertEqual(b, "application/pdf", acceptedTypes[3].spec)
+	utils.AssertEqual(b, "image/png", acceptedTypes[4].spec)
+	utils.AssertEqual(b, "image/jpeg", acceptedTypes[5].spec)
+	utils.AssertEqual(b, "image/gif", acceptedTypes[6].spec)
+	utils.AssertEqual(b, "text/plain", acceptedTypes[7].spec)
+	utils.AssertEqual(b, "image/*", acceptedTypes[8].spec)
+	utils.AssertEqual(b, "text/*", acceptedTypes[9].spec)
+	utils.AssertEqual(b, "*/*", acceptedTypes[10].spec)
+}
+
 // go test -v -run=^$ -bench=Benchmark_App_ETag -benchmem -count=4
 func Benchmark_Utils_ETag(b *testing.B) {
 	app := New()
@@ -219,13 +304,6 @@ func Test_Utils_Parse_Address(t *testing.T) {
 		utils.AssertEqual(t, c.host, host, "addr host")
 		utils.AssertEqual(t, c.port, port, "addr port")
 	}
-}
-
-func Test_Utils_GetOffer(t *testing.T) {
-	t.Parallel()
-	utils.AssertEqual(t, "", getOffer("hello", acceptsOffer))
-	utils.AssertEqual(t, "1", getOffer("", acceptsOffer, "1"))
-	utils.AssertEqual(t, "", getOffer("2", acceptsOffer, "1"))
 }
 
 func Test_Utils_TestConn_Deadline(t *testing.T) {

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -221,7 +221,7 @@ func Test_Utils_Parse_Address(t *testing.T) {
 	}
 }
 
-func Test_Utils_GetOffer(t *testing.T) {
+func Test_GetOffer(t *testing.T) {
 	t.Parallel()
 	utils.AssertEqual(t, "", getOffer("hello", acceptsOffer))
 	utils.AssertEqual(t, "1", getOffer("", acceptsOffer, "1"))

--- a/helpers_test.go
+++ b/helpers_test.go
@@ -72,8 +72,10 @@ func Test_Utils_GetOffer(t *testing.T) {
 	utils.AssertEqual(t, "", getOffer("text/html", acceptsOfferType))
 	utils.AssertEqual(t, "", getOffer("text/html", acceptsOfferType, "application/json"))
 	utils.AssertEqual(t, "", getOffer("text/html;q=0", acceptsOfferType, "text/html"))
+	utils.AssertEqual(t, "", getOffer("application/json, */*; q=0", acceptsOfferType, "image/png"))
 	utils.AssertEqual(t, "application/xml", getOffer("text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", acceptsOfferType, "application/xml", "application/json"))
 	utils.AssertEqual(t, "text/html", getOffer("text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8", acceptsOfferType, "text/html"))
+	utils.AssertEqual(t, "application/pdf", getOffer("text/plain;q=0,application/pdf;q=0.9,*/*;q=0.000", acceptsOfferType, "application/pdf", "application/json"))
 	utils.AssertEqual(t, "application/pdf", getOffer("text/plain;q=0,application/pdf;q=0.9,*/*;q=0.000", acceptsOfferType, "application/pdf", "application/json"))
 
 	utils.AssertEqual(t, "", getOffer("utf-8, iso-8859-1;q=0.5", acceptsOffer))

